lib/pbench/test/unit/agent/test_config.py

```diff
@@ -1,36 +1,233 @@
+        monkeypatch.setattr(
+            MockedConfigParser, "__init__", Mock_MockedConfigParser.__init__
+        )
+        with pytest.raises(BadConfig) as e:
+            PbenchAgentConfig(self.config)
+        assert expected_error_msg in str(e)
+
+    def test_install_dir_exist(self, monkeypatch):
+        """Verify Installation Directory"""
+
+        class Mock_MockedConfigParser:
+            def __init__(self):
+                self._sections = {
+                    "pbench-agent": {
+                        "install-dir": "/mock/install_dir",
+                        "pbench_run": "/mock/pbench-run",
+                        "pbench_log": "/mock/pbench.log",
+                    },
+                    "results": {},
+                }
+
+        expected_error_msg = "BadConfig(\"pbench installation directory, '/mock/install_dir', does not exist\")"
+
+        monkeypatch.setattr(PbenchConfig, "__init__", MockPbenchConfig.__init__)
+        monkeypatch.setattr(
+            MockedConfigParser, "__init__", Mock_MockedConfigParser.__init__
+        )
+        with pytest.raises(BadConfig) as e:
+            PbenchAgentConfig(self.config)
+        assert expected_error_msg in str(e)
+
+    def test_install_dir_directory_creation(self, monkeypatch):
+        """Verify creating Installation Directory"""
+
+        class Mock_MockedConfigParser:
+            def __init__(self):
+                self._sections = {
+                    "pbench-agent": {
+                        "install-dir": "/mock/install_dir",
+                        "pbench_run": "/mock/pbench-run",
+                        "pbench_log": "/mock/pbench.log",
+                    },
+                    "results": {},
+                }
+
+        expected_error_msg = (
+            "BadConfig(\"[Errno 2] No such file or directory: '/mock/pbench-run'\")"
+        )
+
+        monkeypatch.setattr(PbenchConfig, "__init__", MockPbenchConfig.__init__)
+        monkeypatch.setattr(
+            MockedConfigParser, "__init__", Mock_MockedConfigParser.__init__
+        )
+        monkeypatch.setattr(Path, "is_dir", mock_is_dir)
+        with pytest.raises(BadConfig) as e:
+            PbenchAgentConfig(self.config)
+        assert expected_error_msg in str(e)
+
+    def test_valid_PbenchAgentConfig(self, monkeypatch):
+        """Verify Valid Config file"""
+        config = "/mock/pbench.cfg"
+        monkeypatch.setattr(PbenchConfig, "__init__", MockPbenchConfig.__init__)
+        monkeypatch.setattr(Path, "is_dir", mock_is_dir)
+        monkeypatch.setattr(Path, "mkdir", mock_mkdir)
+        PbenchAgentConfig(config)
+
+
+def test_invalid_config(setup):
+    """Verify Invalid Config File"""
+    shutil.copyfile(
+        "./lib/pbench/test/unit/agent/config/pbench-agent.invalid.cfg",
+        str(setup["cfg_dir"] / "pbench-agent.invalid.cfg"),
     )
+    with pytest.raises(BadConfig):
+        PbenchAgentConfig(setup["cfg_dir"] / "pbench-agent.invalid.cfg")
```

Comment:
This looks like a _functional_ test to me -- as such it should probably be marked with a `TODO` comment indicating that it ought to be moved out of here once a suitable home for it exists.

Reply:
I thought of leaving it here for testing an Invalid config file but let's get rid of it :)

Reply:
This is kinda on the edge; this is something that *could* be tested using the external CLI interfaces, but it's more easily tested with the internal `PbenchAgentConfig` constructor, and as that's not a supported external interface it's arguably more of a unit test even though this isn't technically mocking the dependencies.

Reply:
I think that this is a good and appropriate test; I just don't think it belongs here (but, perhaps it is OK for it to live here until a better home can be found for it).

Whether the CUT is tested via external or internal interfaces is (IMO) orthogonal to the question of whether the test is a _functional_ test or a _unit_ test.  What really distinguishes a unit test is the fact that it tests the "unit" in a way which isolates the CUT from its dependencies.  The operation of this test relies on the file system, so it is clearly **not** a unit test...so, I think that makes it a functional test, by default.  😀